master/internal/stream/supervisor.go

```diff
@@ -0,0 +1,150 @@
+package stream
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+	"github.com/gorilla/websocket"
+	"github.com/labstack/echo/v4"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/determined-ai/determined/master/pkg/syncx/errgroupx"
+)
+
+const (
+	minInterval = time.Second
+	maxInterval = time.Minute
+	maxRetries  = 5
+)
+
+// Supervisor manages the context for underlying PublisherSet.
+type Supervisor struct {
+	cond *sync.Cond
+	// these things change each restart of the PublisherSet
+	subctx context.Context
+	ps     *PublisherSet
+}
+
+// NewSupervisor creates a new Supervisor.
+func NewSupervisor() *Supervisor {
+	var lock sync.Mutex
+	return &Supervisor{cond: sync.NewCond(&lock)}
+}
+
+// monitorPermissionChanges listens for permission changes, updates the PublisherSet
+// to signal to boot streamers, returns an error in the event of a failure to listen.
+func monitorPermissionChanges(ctx context.Context, ps *PublisherSet) error {
```

Comment:
this is just a nit and is totally optional: I like that this isn't a method on the PublisherSet, but it still feels closer to the publisher set than to the supervisor.  Or really... it feels closer to the publishLoop than it feels to the supervisor.  So it might fit better in system.go.

Or mabye system.go is getting big enough to be broken up.  That's also possible.